modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/planner/JoinCommutePlannerTest.java

```diff
@@ -76,6 +82,100 @@ public class JoinCommutePlannerTest extends AbstractPlannerTest {
         );
     }
 
+    /** Ensures that join commute rules are disabled if joins count is too high. Checks the tables order in the query. */
+    @Test
+    public void testCommuteDisabledForManyJoins() throws Exception {
+        IgniteTypeFactory f = new IgniteTypeFactory(IgniteTypeSystem.INSTANCE);
+
+        int maxJoinsToOptimize = Math.max(PlannerHelper.MAX_JOINS_TO_COMMUTE_INPUTS, PlannerHelper.MAX_JOINS_TO_COMMUTE);
+
+        int tablesCnt = maxJoinsToOptimize + 3;
+
+        try {
+            for (int i = 0; i < tablesCnt; ++i) {
+                publicSchema.addTable(
+                    "TBL" + i,
+                    new TestTable(
+                        new RelDataTypeFactory.Builder(f)
+                            .add("ID", f.createJavaType(Integer.class))
+                            .build(), Math.pow(100, (1 + (i % 3)))) {
+
+                        @Override public IgniteDistribution distribution() {
+                            return IgniteDistributions.affinity(0, "TEST_CACHE", "hash");
+                        }
+                    }
+                );
+            }
+
+            doTestCommuteDisabledForManyJoins(maxJoinsToOptimize + 1, false, "TBL");
+
+            doTestCommuteDisabledForManyJoins(maxJoinsToOptimize + 2, true, "TBL");
```

Comment:
`maxJoinsToOptimize + 2` looks confusing. Lets use the same joinsCnt and add subquery to "select" variable when withCorrelated is true (outside of for-loop).
Also, we do not check correlate when joinsCnt = maxJoinsToOptimize (if we want to check that correlate is not counted as join), so check with maxJoinsToOptimize + 1 joins looks meaningless, since we have checked query with the same count of joins before.